src/nvim/highlight_group.c

```diff
@@ -1521,6 +1522,228 @@ static void highlight_list_one(const int id)
+    hl = arena_dict(arena, 1);
+    char *link = hl_table[sgp->sg_link - 1].sg_name;
+    PUT_C(hl, "link", STRING_OBJ(cstr_as_string(link)));
+  } else {
+    hl = arena_dict(arena, HLATTRS_DICT_SIZE);
+    HlAttrs ae = syn_attr2entry(sgp->sg_attr);
+    int mask = ae.rgb_ae_attr;
+
+    // Ordered by `:h nvim_set_hl`
+    char hexbuf[8];
+    if (ae.rgb_fg_color != -1) {
+      if (use_hex) {
+        PUT_C(hl, "fg",
+              STRING_OBJ(cstr_as_string(xstrdup(coloridx_to_name(sgp->sg_rgb_fg_idx, sgp->sg_rgb_fg,
+                                                                 hexbuf)))));
+      } else {
+        PUT_C(hl, "fg", INTEGER_OBJ(ae.rgb_fg_color));
+      }
+    }
+
+    if (ae.rgb_bg_color != -1) {
+      if (use_hex) {
+        PUT_C(hl, "bg",
+              STRING_OBJ(cstr_as_string(xstrdup(coloridx_to_name(sgp->sg_rgb_bg_idx, sgp->sg_rgb_bg,
+                                                                 hexbuf)))));
+      } else {
+        PUT_C(hl, "bg", INTEGER_OBJ(ae.rgb_bg_color));
+      }
+    }
+
+    if (ae.rgb_sp_color != -1) {
+      if (use_hex) {
+        PUT_C(hl, "sp",
+              STRING_OBJ(cstr_as_string(xstrdup(coloridx_to_name(sgp->sg_rgb_sp_idx, sgp->sg_rgb_sp,
+                                                                 hexbuf)))));
+      } else {
+        PUT_C(hl, "sp", INTEGER_OBJ(ae.rgb_sp_color));
+      }
+    }
+
+    if (ae.hl_blend > -1) {
+      PUT_C(hl, "blend", INTEGER_OBJ(ae.hl_blend));
+    }
+
+    if (mask & HL_BOLD) {
+      PUT_C(hl, "bold", BOOLEAN_OBJ(true));
+    }
+
+    if (mask & HL_STANDOUT) {
+      PUT_C(hl, "standout", BOOLEAN_OBJ(true));
+    }
+
+    switch (mask & HL_UNDERLINE_MASK) {
+    case HL_UNDERLINE:
+      PUT_C(hl, "underline", BOOLEAN_OBJ(true));
+      break;
+
+    case HL_UNDERCURL:
+      PUT_C(hl, "undercurl", BOOLEAN_OBJ(true));
+      break;
+
+    case HL_UNDERDOUBLE:
+      PUT_C(hl, "underdouble", BOOLEAN_OBJ(true));
+      break;
+
+    case HL_UNDERDOTTED:
+      PUT_C(hl, "underdotted", BOOLEAN_OBJ(true));
+      break;
+
+    case HL_UNDERDASHED:
+      PUT_C(hl, "underdashed", BOOLEAN_OBJ(true));
+      break;
+    }
+
+    if (mask & HL_STRIKETHROUGH) {
+      PUT_C(hl, "strikethrough", BOOLEAN_OBJ(true));
+    }
+
+    if (mask & HL_ITALIC) {
+      PUT_C(hl, "italic", BOOLEAN_OBJ(true));
+    }
+
+    if (mask & HL_INVERSE) {
+      PUT_C(hl, "reverse", BOOLEAN_OBJ(true));
+    }
+
+    if (mask & HL_NOCOMBINE) {
+      PUT_C(hl, "nocombine", BOOLEAN_OBJ(true));
+    }
+
+    mask = ae.cterm_ae_attr;
+
+    if (ae.cterm_fg_color != 0) {
+      PUT_C(hl, "ctermfg", INTEGER_OBJ(ae.cterm_fg_color - 1));
+    }
+
+    if (ae.cterm_bg_color != 0) {
+      PUT_C(hl, "ctermbg", INTEGER_OBJ(ae.cterm_bg_color - 1));
+    }
+
+    if (!(mask & HL_UNDERLINE_MASK) && ae.cterm_ae_attr != 0) {
+      // Ordered by `:h highlight-cterm`
+      Dictionary cterm = arena_dict(arena, HLATTRS_DICT_SIZE);
+      mask = ae.cterm_ae_attr;
```

Comment:
This doesn't look right:
1. `mask` is assigned `ae.cterm_ae_attr` again when it is already assigned that value.
2. `mask & HL_UNDERLINE_MASK` can only be 0 inside this branch.

Reply:
1. Thanks, I will fix that
2. Try this
```
:hi String cterm=italic,bold
:lua=vim.api.nvim_get_hl(0, { name = "String" })
{
  cterm = {
    bold = true,
    italic = true
  },
  fg = "#a6e3a1"
}
```

`mask & HL_UNDERLINE_MASK` isn't 0 in this case.

Reply:
Resolved after the `hlattrs2dict` refactor.